### ISSUE TITLE
What Are the Training Time and Space Complexities for Mamba vs. Traditional Transformers?

### ISSUE THREAD

**Issue by mrconter1** [state-spaces/mamba#196]
Hi!

I've been exploring the Mamba architecture with great interest, especially its computational efficiencies compared to traditional Transformer models. The selective state space approach and the potential for handling long sequences efficiently are particularly intriguing.

However, I would appreciate some clarification on the exact time and space complexities associated with different operations within Mamba compared to Transformers. Specifically, I'm interested in understanding the complexities related to training, autoregressive inference, and the memory requirements for maintaining the model's state.

From the available documentation and discussions, I've gathered that:

* Training complexity for Mamba is O(L), assuming L is the sequence length, which contrasts with the O(L^2) complexity typical of Transformer models due to their quadratic scaling with sequence length.
* Autoregressive inference complexity in Mamba is indicated as O(T) per step, where T is a fixed constant time for compression and selection of relevant information from the input, differing from Transformers that exhibit an O(L) complexity per step, increasing linearly with the sequence length.
* Memory complexity for Mamba is described as O(C), with C representing the fixed size of the SSM latent state. This is in contrast to the O(L) memory complexity of Transformers, where memory requirements scale linearly with sequence length.

Are this correct? Could you provide more detailed insights into how these complexities are derived and any underlying assumptions?

Thank you for your pioneering work on Mamba, and I look forward to your insights and further clarifications on these questions.

Best regards

**Comment by albertfgu** [comment]
Yes, these are correct. Please check out the paper or the many blogs/videos/tutorials on the topic for more details.
